### PR TITLE
switch to binary Compiler

### PR DESCRIPTION
**Pull request by zhfnjust**

### CODE REVIEW

**Comment by zhfnjust** [comment]
ci failed because binary compiler can't run on linux. just update compiler_dist, no need to change code

File: .travis.yml [hunk 1]
```diff
@@ -10,12 +10,13 @@ before_install:
...
   - npm link
-  - git clone -b array https://github.com/scrypt-sv/boilerplate.git
+  - git clone -b master https://github.com/scrypt-sv/boilerplate.git
   - cd boilerplate
   - npm install
   - npm link scryptlib  # use this version
   #- npm ls scryptlib    # double check the latest version is used
-  - npm test
\ No newline at end of file
+  - npm test -- --ci=true
```
Comment (xhliu) [inline .travis.yml:22]: No need to change

File: src/compilerWrapper.ts [hunk 2]
```diff
@@ -106,9 +107,10 @@ export function compile(
 	const outputFiles = {};
 	try {
 		const sourceContent = source.content !== undefined ? source.content : readFileSync(sourcePath, 'utf8');
-		const cmdPrefix = settings.cmdPrefix || `npx ${npxArg} scryptc${settings.scVersion ? '@' + settings.scVersion : ''}`;
+		const cmdPrefix = settings.cmdPrefix || getDefaultsCryptc();
 		const cmd = `${cmdPrefix} compile ${settings.asm || settings.desc ? '--asm' : ''} ${settings.ast || settings.desc ? '--ast' : ''} ${settings.debug == false ? '' : '--debug'} -r -o "${outputDir}" ${settings.cmdArgs ? settings.cmdArgs : ''}`;
-		const output = execSync(cmd, { input: sourceContent, cwd: curWorkingDir }).toString();
+		let output = execSync(cmd, { input: sourceContent, cwd: curWorkingDir }).toString();
+		output = output.split(/\r?\n/g).join('\n');
```
Comment (xhliu) [inline src/compilerWrapper.ts:113]: add a comment explaining why this is needed

File: src/compilerWrapper.ts [hunk 3]
```diff
@@ -368,4 +370,47 @@ function getABIDeclaration(astRoot): { contract: string, abi: Array<ABIEntity> }
 		contract: mainContract['name'],
 		abi: interfaces
 	};
-}
\ No newline at end of file
+}
+
+export function getPlatformsCryptc() : string {
```
Comment (xhliu) [inline src/compilerWrapper.ts:375]: rename to `getPlatformScryptc`

File: src/compilerWrapper.ts [hunk 4]
```diff
@@ -368,4 +370,47 @@ function getABIDeclaration(astRoot): { contract: string, abi: Array<ABIEntity> }
...
+	switch (os.platform()) {
+		case "win32":
+			return "compiler/scryptc/win32/scryptc.exe";
+		case "linux":
+			return "compiler/scryptc/linux/scryptc";
+		case "darwin":
+			return "compiler/scryptc/mac/scryptc";
+		default:
+			throw "sCrypt don't support your OS now";
```
Comment (xhliu) [inline src/compilerWrapper.ts:384]: `throw `sCrypt doesn't support ${os.platform()}`;`

File: src/compilerWrapper.ts [hunk 5]
```diff
@@ -368,4 +370,47 @@ function getABIDeclaration(astRoot): { contract: string, abi: Array<ABIEntity> }
...
+	}
+}
+
+
+
+export function getDefaultsCryptc(): string {
```
Comment (xhliu) [inline src/compilerWrapper.ts:390]: `getDefaultScryptc`

File: src/compilerWrapper.ts [hunk 6]
```diff
@@ -368,4 +370,47 @@ function getABIDeclaration(astRoot): { contract: string, abi: Array<ABIEntity> }
...
+	const homedir = os.homedir();
+	const extensions =  join(homedir, ".vscode/extensions");
+	if(!existsSync(extensions)) {
+		throw `No vscode extensions found, Please check if vscode is installed on your machine.`
```
Comment (xhliu) [inline src/compilerWrapper.ts:394]: throw `No Visual Studio Code extensions found. Please ensure Visual Studio Code is installed.`

File: src/compilerWrapper.ts [hunk 7]
```diff
@@ -368,4 +370,47 @@ function getABIDeclaration(astRoot): { contract: string, abi: Array<ABIEntity> }
...
+		default:
+			throw "sCrypt don't support your OS now";
+	}
+}
+
+
+
+export function getDefaultsCryptc(): string {
+	const homedir = os.homedir();
+	const extensions =  join(homedir, ".vscode/extensions");
```
Comment (xhliu) [inline src/compilerWrapper.ts:392]: extensions -> extensionDir

File: src/compilerWrapper.ts [hunk 8]
```diff
@@ -368,4 +370,47 @@ function getABIDeclaration(astRoot): { contract: string, abi: Array<ABIEntity> }
...
+	if(!existsSync(extensions)) {
+		throw `No vscode extensions found, Please check if vscode is installed on your machine.`
+	}
+
+    let sCrypt = readdirSync(extensions).filter(dir => {
```
Comment (xhliu) [inline src/compilerWrapper.ts:397]: Replace `filter` w/ [some](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/some).

sCrypt -> bool scryptExist

File: src/compilerWrapper.ts [hunk 9]
```diff
@@ -368,4 +370,47 @@ function getABIDeclaration(astRoot): { contract: string, abi: Array<ABIEntity> }
...
+		if(dir.indexOf("bsv-scrypt.scrypt-") > -1 ) {
+			return true;
+		} 
+		return false;
+	});
+
+	if(sCrypt.length == 0) {
+		throw `No sCrypt IDE found, Please install sCrypt IDE at vscode extensions marketplace: https://marketplace.visualstudio.com/items?itemName=bsv-scrypt.sCrypt`;
```
Comment (xhliu) [inline src/compilerWrapper.ts:405]: throw `No sCrypt extension found. Please install it at extension marketplace: https://marketplace.visualstudio.com/items?itemName=bsv-scrypt.sCrypt`;

File: src/compilerWrapper.ts [hunk 10]
```diff
@@ -368,4 +370,47 @@ function getABIDeclaration(astRoot): { contract: string, abi: Array<ABIEntity> }
...
+	} 
+
+	let scryptc = join(extensions, sCrypt[0], getPlatformsCryptc());
+
+	if(!existsSync(scryptc)) {
+		throw `No scryptc found, Please update your sCrypt extension to latest version`;
```
Comment (xhliu) [inline src/compilerWrapper.ts:411]: throw `No sCrypt compiler found. Please update your sCrypt extension to the latest version`;

File: test/compilerWrapper.test.ts [hunk 11]
```diff
@@ -16,12 +15,12 @@ describe('compile()', () => {
...
     const result = compileContract("p2pkh_wrong.scrypt", "fixture/invalid");
-
+    
     assert.isAbove(result.errors.length, 0, "exist Errors");
     assert.include(result.errors[0].type, 'SyntaxError', 'contract has SyntaxError');
   })
-
-  it('should generate description file properly', () => {
+  
+it('should generate description file properly', () => {
```
Comment (xhliu) [inline test/compilerWrapper.test.ts:23]: Misaligned with previous `it`.

File: test/helper.ts [hunk 12]
```diff
@@ -20,3 +20,18 @@ export function newTx(inputSatoshis: number) {
   };
   return new bsv.Transaction().from(utxo);
 }
+
+
+export function compileContract(fileName: string, folder: string): CompileResult {
+
+
```
Comment (xhliu) [inline test/helper.ts:27]: Remove blank lines.

**Review by xhliu — CHANGES_REQUESTED** [review]
Look good overall 👍 

Just some minor refactors.

File: .travis.yml (hunk 1 above)
Comment (zhfnjust) [reply .travis.yml:22]: we need to get scryptc while run test on ci. see this commit
https://github.com/sCrypt-Inc/boilerplate/commit/be138595c611f4ec0a7d7346c7c88cf601b3bf48

File: src/compilerWrapper.ts (hunk 2 above)
Comment (zhfnjust) [reply src/compilerWrapper.ts:113]: added

File: src/compilerWrapper.ts (hunk 3 above)
Comment (zhfnjust) [reply src/compilerWrapper.ts:375]: done

File: src/compilerWrapper.ts (hunk 4 above)
Comment (zhfnjust) [reply src/compilerWrapper.ts:384]: done

File: src/compilerWrapper.ts (hunk 8 above)
Comment (zhfnjust) [reply src/compilerWrapper.ts]: see fd7ae462eb523699dc772456c293c322d9339378

File: src/compilerWrapper.ts (hunk 9 above)
Comment (zhfnjust) [reply src/compilerWrapper.ts]: see fd7ae462eb523699dc772456c293c322d9339378

File: src/compilerWrapper.ts (hunk 2 above)
Comment (zhfnjust) [reply src/compilerWrapper.ts:114]: see d4172b71e62e4f3b68a150aea78babf5e54f3d39

File: src/compilerWrapper.ts (hunk 3 above)
Comment (zhfnjust) [reply src/compilerWrapper.ts]: done

File: src/compilerWrapper.ts (hunk 4 above)
Comment (zhfnjust) [reply src/compilerWrapper.ts]: done

**Pull request merged**